node/actors/bft/src/replica/leader_prepare.rs

```diff
@@ -266,6 +269,18 @@ impl StateMachine {
                         header: message.proposal,
                     });
                 }
+
+                // Payload should be valid.
+                if let Err(err) = self
+                    .storage
+                    .verify_payload(ctx, message.proposal.number, payload)
+                    .await
```

Comment:
Why delegate the verification to the storage? I think it would make more sense to have this method as part of the PayloadSource trait.

Reply:
PayloadSource is consumed by leader, verify_payload is consumed by replica. verify_payload() should be thought of a dryrun of put_block(), hence the placement. We expect the blockchains to be stateful, so verify_payload() will need to have access to the state (storage) anyway.

Reply:
I'm not saying that this API perfect though. It's just the best version I've come up with for now.